Solver: Improve handling of failures involving whole packages, such as unknown packages.

grayjay opened the pull request:
Two main commits:

#### Solver: Represent an unknown package with a failure node.

Previously, the solver represented an unknown package with an empty PChoice
node, because an unknown package is like a package with no versions.  Now that
--reject-unconstrained-dependencies also needs to reject whole packages, it is
simpler to handle both types of failures similarly, and use a failure node for
both.  Another advantage of using a failure node now is that the node can store
more information about why the package was rejected.  The explicit node also
simplifies generating the solver log.

This commit treats a failure node under a goal choice as a backjump in
D.S.Modular.Explore.exploreLog, which fixes #5502.

#### Solver: Print package rejection message on a single line.

The one-line error message is more similar to the message for rejecting a
package version.

Closes #5501.

Before:

```
[__0] trying: A-3.0.0 (user goal)
[__1] next goal: C (dependency of A)
[__1] fail (not a user-provided goal nor mentioned as a constraint, but reject-unconstrained-dependencies was set)
[__1] fail (backjumping, conflict set: A, C)
```

After:

```
[__0] trying: A-3.0.0 (user goal)
[__1] rejecting: C (not a user-provided goal nor mentioned as a constraint, but reject-unconstrained-dependencies was set) (dependency of A)
[__1] fail (backjumping, conflict set: A, C)
```

This commit also expands the tests for the error message for
--reject-unconstrained-dependencies.

_______________________________________________________

I'm not sure that putting the rejected package message on one line is an improvement, now that I've seen it, since that message contains more information than the rejected version message.  Let me know if you have any suggestions for improving it or prefer the original format.

/cc @kosmikus @quasicomputational 

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [ ] Any changes that could be relevant to users have been recorded in the changelog.
* [ ] The documentation has been updated, if necessary.
* [ ] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

quasicomputational commented:
Yeah, I'm also not sure about the one-liner. The `(dependency of A)` information is important, and I think we should put it at the front of the line, or on a line itself. Having it dangling at the end like that is sort of the worst of all worlds.

Here's a sketch of what it could look like with it at the front of the line:

```
[__0] trying: A-3.0.0 (user goal)
[__1] rejecting: C (dependency of A; not a user-provided goal nor mentioned as a constraint, but reject-unconstrained-dependencies was set)
[__1] fail (backjumping, conflict set: A, C)
```

Or, with longer package names:

```
[__0] trying: some-package-3.0.0 (user goal)
[__1] rejecting: some-other-package (dependency of some-package; not a user-provided goal nor mentioned as a constraint, but reject-unconstrained-dependencies was set)
[__1] fail (backjumping, conflict set: some-package, some-other-package)
```

That bikeshedding aside, the code looks sensible.

23Skidoo commented:
Code looks good from what I can tell. Re: the error message change, I think I prefer the original format.

grayjay commented:
Thanks for the feedback!  I removed the error message change and updated the unit test to test the original message.  I was worried that putting "dependency of A" between "rejecting: C" and the reason for the failure could imply that "dependency of A" was part of the reason that C was rejected.
